### ISSUE TITLE
Cluster monitoring cannot be restored once disabled

### ISSUE THREAD

**Issue by materemias** [rancher/rancher#19209]
**What kind of request is this (question/bug/enhancement/feature request):**
bug

**Steps to reproduce (least amount of steps as possible):**
- enable cluster monitoring
- disable cluster monitoring
- delete cattle-prometheus namespace and leftover apps
- re-enable cluster monitoring

**Result:**
only cluster-monitoring app is deployed, cluster-alerting and monitoring-operator doesn't get deployed, leaving cluster monitoring in a half-ready state

**Other details that may be helpful:**

**Environment information**
- Rancher version (`rancher/rancher`/`rancher/server` image tag or shown bottom left in the UI): v2.2.0
- Installation option (single install/HA): single

**Cluster information**
- Cluster type (Hosted/Infrastructure Provider/Custom/Imported): GKE
- Machine type (cloud/VM/metal) and specifications (CPU/memory): cloud
- Kubernetes version (use `kubectl version`):

```
(paste the output here)
```

- Docker version (use `docker version`):

```
(paste the output here)
```

**Comment by vincent927** [comment]
same problem

**Comment by hmonadjem** [comment]
same problem, is there a way to create the missing apps manually?

**Comment by loganhz** [comment]
Can you try to disable the monitoring and re-enable it again?

**Comment by thxCode** [comment]
@materemias , `monitoring-operator` is a crucial component app to schedule Prometheus and Alertmanager. That means if you drop it during enabling cluster/project Monitoring or Alerting, you need to rollback both Monitoring and Alerting to the disabling status as a workaround. 

For example: You can remove those `AlertGroup` which have already connected with a `Notifier`, and re-enable Monitoring. 

@loganhz , maybe we can hide those Apps from UI to prevent some undesired deletion behaviors?

**Comment by materemias** [comment]
@thxCode the issue is that monitor-operator doesn't event get deployed, only these: 
![image](https://user-images.githubusercontent.com/529994/55276399-a927d000-52f3-11e9-8649-48d0358d3e1d.png)
I did not drop it during enabling cluster monitoring, I disabled cluster monitoring and since it had some leftover I deleted the namespace. Now I cannot get it up and running again.

**Comment by thxCode** [comment]
@materemias , Yes, I know you just want to clean up the scene. I'm going to explain those leftover. The following matrix shows the relationships for Monitoring/Alerting Apps.

| Rancher Tool | Functional App | Manage App |
|---|---|---|
| Cluster Monitoring | `System` Project / `cluster-monitoring` App | `System` Project / `monitoring-operator` App |
| Project Monitoring | Pre Project / `project-monitoring` App | `System` Project / `monitoring-operator` App |
| Alerting | `System` Project / `cluster-alerting` App | `System` Project / `monitoring-operator` App |

There is still living a `monitoring-operator` App after disabling Cluster Monitoring because Project Monitoring or Alerting is enabling.

We think this is still a bug, we need to make the System Apps more stronger.

**Comment by materemias** [comment]
And is there a way I can get monitoring-operator back for now? (It is not listed under apps)

**Comment by materemias** [comment]
@thxCode I have disabled all monitoring and inactivated all alerting in the cluster which is affected and tried to start over as you suggested but when I enable cluster monitoring I end up with the same.
Rancher logs show this error
```
[ERROR] ClusterController c-th9ct [cluster-monitoring-handler] failed with : unable to sync Cluster c-th9ct(asdasd): failed to detect the installation status of monitoring components: Prometheus StatefulSet isn't deployed
```

**Comment by alena1108** [comment]
@loganhz for 2.2.2, lets make sure that:

1) Following use case works:

* Enable monitoring when there is not enough resources.
* Disable monitoring.
* Add more resources and enable monitoring.

2) For the case when manual removal is done for one of the monitoring components, we should provide the workaround for 2.2.2, and fix it in 2.2.3

**Comment by doog33k** [comment]
I have exactly the same  behavior as @materemias , unfortunately, once the monitoring-operator is deleted , it is impossible to install it again, even manually :'(
I can't find a way to install it again

**Comment by loganhz** [comment]
@doog33k @materemias @hmonadjem @vincent927 

We will fix it in 2.2.2.

If you hit this issue in 2.2.0 or 2.2.1, currently the workaround is not easy. 
But it can get your cluster monitoring working again.

Workaround in 2.2.0 and 2.2.1 if you manually delete `cattle-prometheus` namespace.

1.  Disable cluster monitoring
2. Go to System project -> Apps
3. Remove `cluster-monitoring` and `monitoring-operator` apps if the apps cannot be deleted.
4. docker exec into your rancher server container
5. `kubectl get cluster` to find your cluster id
6. `kubectl edit cluster ${cluster_id}`
7.  Delete the `PrometheusOperatorDeployed` condition and save the change.
![image](https://user-images.githubusercontent.com/21168270/55466854-a0583680-5632-11e9-9faa-e5fc2f6b8310.png)
8. Enable the cluster monitoring again

**Comment by doog33k** [comment]
Thanks @loganhz  for your answer.
I tried to use your workarround, but unfortunately I was unable to install vi/vim, I had connection timeout during installation of the package inside the container, i may have to configure a proxy because I'm on a private cloud.

**Comment by hmonadjem** [comment]
@loganhz 
thanks, the workaround started all apps needed but I now run into a new issue:

App: cluster-monitoring
Workload: prometheus-cluster-monitoring
Container: prometheus

failes to start with the following logs:
```
level=warn ts=2019-04-03T12:27:40.278783148Z caller=main.go:295 deprecation_notice="\"storage.tsdb.retention\" flag is deprecated use \"storage.tsdb.retention.time\" instead."
level=info ts=2019-04-03T12:27:40.278856671Z caller=main.go:302 msg="Starting Prometheus" version="(version=2.7.1, branch=HEAD, revision=62e591f928ddf6b3468308b7ac1de1c63aa7fcf3)"
level=info ts=2019-04-03T12:27:40.278881428Z caller=main.go:303 build_context="(go=go1.11.5, user=root@f9f82868fc43, date=20190131-11:16:59)"
level=info ts=2019-04-03T12:27:40.278908205Z caller=main.go:304 host_details="(Linux 4.4.0-1052-aws #61-Ubuntu SMP Mon Feb 12 23:05:58 UTC 2018 x86_64 prometheus-cluster-monitoring-0 (none))"
level=info ts=2019-04-03T12:27:40.278934985Z caller=main.go:305 fd_limits="(soft=1048576, hard=1048576)"
level=info ts=2019-04-03T12:27:40.27895894Z caller=main.go:306 vm_limits="(soft=unlimited, hard=unlimited)"
level=info ts=2019-04-03T12:27:40.279734659Z caller=main.go:620 msg="Starting TSDB ..."
level=info ts=2019-04-03T12:27:40.27978457Z caller=web.go:416 component=web msg="Start listening for connections" address=127.0.0.1:9090
level=warn ts=2019-04-03T12:27:40.280110658Z caller=wal.go:116 component=tsdb msg="last page of the wal is torn, filling it with zeros" segment=/prometheus/wal/00000000
level=info ts=2019-04-03T12:27:43.090615857Z caller=main.go:635 msg="TSDB started"
level=info ts=2019-04-03T12:27:43.090680387Z caller=main.go:695 msg="Loading configuration file" filename=/etc/prometheus/config_out/prometheus.env.yaml
level=info ts=2019-04-03T12:27:43.176338647Z caller=kubernetes.go:201 component="discovery manager scrape" discovery=k8s msg="Using pod service account via in-cluster config"
level=info ts=2019-04-03T12:27:43.177736039Z caller=kubernetes.go:201 component="discovery manager scrape" discovery=k8s msg="Using pod service account via in-cluster config"
level=info ts=2019-04-03T12:27:43.178533836Z caller=kubernetes.go:201 component="discovery manager scrape" discovery=k8s msg="Using pod service account via in-cluster config"
level=info ts=2019-04-03T12:27:43.179318403Z caller=kubernetes.go:201 component="discovery manager scrape" discovery=k8s msg="Using pod service account via in-cluster config"
level=info ts=2019-04-03T12:27:43.183250495Z caller=main.go:722 msg="Completed loading of configuration file" filename=/etc/prometheus/config_out/prometheus.env.yaml
level=info ts=2019-04-03T12:27:43.183289373Z caller=main.go:589 msg="Server is ready to receive web requests."
level=error ts=2019-04-03T12:27:45.374133789Z caller=notifier.go:481 component=notifier alertmanager=http://alertmanager-operated.cattle-prometheus:9093/api/v1/alerts count=0 msg="Error sending alert" err="Post http://alertmanager-operated.cattle-prometheus:9093/api/v1/alerts: dial tcp: lookup alertmanager-operated.cattle-prometheus on 10.43.0.10:53: no such host"
```


![image](https://user-images.githubusercontent.com/4346277/55478759-bf4bce00-561c-11e9-9209-977affce986b.png)
![Screenshot 2019-04-03 at 14 28 46](https://user-images.githubusercontent.com/4346277/55478816-e30f1400-561c-11e9-84ae-247d0092a4db.png)

Should I create a new issue for this or is this directly related?

**Comment by loganhz** [comment]
@hmonadjem the error message should be unrelated. 
Please go to enable monitoring page and update the Prometheus Memory Limit.
Now it is 500Mib I think. Please change it to 1000Mib. And click save.

**Comment by hmonadjem** [comment]
@loganhz thanks, now it worked. but 1000 was not enough ... 2.4GB are in usage and I did put a limit of 4GB ;)

**Comment by loganhz** [comment]
@hmonadjem could you describe your cluster a bit? like how many nodes, where are they and what's the size and OS for the nodes, please?

**Comment by hmonadjem** [comment]
@loganhz sure, happy to share.

We have 13 Nodes:
10 as Control Plane & Worker (AWS - m4.2xlarge)
3 as etcd (AWS - t2.small)
OS: Ubuntu 16.04.4 LTS (4.4.0-1052-aws)

We have all together 80 CPU and 313 GiB of Memory about 40% of the Memory is reserved and about 80% of the CPU is also reserved and running about 30 Apps with each about 18 Pods in average.

**Comment by loganhz** [comment]
Thanks @hmonadjem . Just FYI, https://github.com/rancher/rancher/issues/19187#issuecomment-478177608 we need to open the node exporter open to see the metrics for nodes

**Comment by materemias** [comment]
The manual fix worked, thanks

**Comment by vincent927** [comment]
@loganhz thanks, it worked

**Comment by alena1108** [comment]
Per @loganhz regular enable->disable->enable monitoring works. The issue happens only when operator is removed manually. Given the workaround, moving out of 2.2.2. Fix to address manual removal scenario will be done in subsequent maintenance release.

**Comment by doog33k** [comment]
Excellent, thanks for your help.

**Comment by doog33k** [comment]
Just to provide a quick update.
I took the time to use the workaround explained by @loganhz 
It worked perfectly.
I would like to thank you for your wonderfull job arroudn Rancher @alena1108 @loganhz 
The v2.2 seems realy terrific

**Comment by deniseschannon** [comment]
@thxCode @guangbochen please review the comments from this issue that might indicate why it isn't working: https://github.com/rancher/rancher/issues/19559

This might not be monitoring specific but something to do with our version of helm?

**Comment by deniseschannon** [comment]
@loganhz could you update this issue title cause the last comments seem to indicate that there is a specific use case to hit this

**Comment by vitobotta** [comment]
Had this and the manual fix worked, thanks!

**Comment by mholasek** [comment]
I have the same problem. Version 2.2.2. No way to bring back the monitoring-operator on a specific cluster (rancher created cluster). When I enable monitoring, it installs the cluster-monitoring app but no monitoring-operator.  After it is deployed I disabled it and the app is still there. If I delete the app and enable once again, same issue. Is there any way to reenable it in Version 2.2.2 or is a fix in 2.2.3 on the way?

**Comment by mholasek** [comment]
Got it. I've edited my HA rancher cluster with kubectl (kubectl edit cluster {clusterid}) and removed the three lines as mentioned by loganhz and now after enabling monitoring the operator will be deployed again. If I disable monitoring both apps will be removed automatically. Thx for the workaround.

**Comment by loganhz** [comment]
Waiting for https://github.com/rancher/rancher/pull/19364 merged and backport to 2.2 branch

**Comment by jiaqiluo** [comment]
@loganhz @thxCode  I cannot get the issue reproduced on my setup. Could you provide more instructions for either reproduction or bug-validation?

**Comment by thxCode** [comment]
@jiaqiluo , reproducing follows as below on v2.2.2:
- enable monitoring
- add a notifier and bind to an alerting group
- you can see a new alertmanager workload in `cattle-prometheus`  namespace
- disable monitoring
- you can see two workloads still keep in `cattle-prometheus` namespace: `alertmanager`, `operator`
- drop `cattle-promethues` namespace

**Comment by jiaqiluo** [comment]
@thxCode Thank you for the step, and I got the issue to be reproduced on v2.2.2

-----

Here is what happens on both Rancher: master and v2.2.3-rc7 

There are 3 apps deployed in total 
 
<img width="1404" alt="screenshot" src="https://user-images.githubusercontent.com/6218999/57399935-27934f80-7187-11e9-9c25-38378cc35d47.png">

Now disable the cluster monitoring, there are still two apps left as the screenshot below shows.
<img width="1448" alt="screenshot" src="https://user-images.githubusercontent.com/6218999/57400074-822cab80-7187-11e9-9cf2-2f43c305cde0.png">

Do we expect to have those 2 apps left? 


Moreover, If I delete those two apps and the namespace `cattle-prometheus` manually, and then enable the cluster monitoring again, the `cluster-alerting` app will NOT be deployed even the notifier and alters settings are unchanged. 
<img width="1483" alt="screenshot" src="https://user-images.githubusercontent.com/6218999/57400200-d20b7280-7187-11e9-82a2-94bfca662c7b.png">

**Comment by jiaqiluo** [comment]
The issue is reopened based on the following observation:

After disabling the cluster monitoring and only the `cluster-monitoring` app being deleted,  if I go to edit any alter related to the metrics in the cluster alerts section, the following error message will show up:
<img width="1665" alt="screenshot" src="https://user-images.githubusercontent.com/6218999/57412033-32101200-71a4-11e9-8b9a-875cb59a3cb2.png">

It indicates that metric-related alters are not working now because the cluster monitoring has been disabled. In this case, should we still keep the `cluster-alerting` and `monitoring-operator` app around? 

cc @loganhz @thxCode

**Comment by loganhz** [comment]
> Now disable the cluster monitoring, there are still two apps left as the screenshot below shows.

Yes, it is expected. Alerting requires cluster-alerting app and operator app

> enable the cluster monitoring again, the cluster-alerting app will NOT be deployed even the notifier and alters settings are unchanged.

This is also expected. If users did the remove manually, users need to click `save` for any alert to trigger it. Monitoring and alerting are 2 different features. So it won't impact each other. But they are sharing the same operator.

> After disabling the cluster monitoring and only the cluster-monitoring app being deleted, if I go to edit any alter related to the metrics in the cluster alerts section, the following error message will show up:

We have some pre-defined expression type alerts which users can always edit it even without turning on monitoring. If you create a new alert after disabling the cluster monitoring, you will not be able to select `expression`.

**Comment by jiaqiluo** [comment]
@loganhz  

> We have some pre-defined expression type alerts which users can always edit it even without turning on monitoring. If you create a new alert after disabling the cluster monitoring, you will not be able to select expression.

I tested the built-in alerts but saw errors. Here is what I did on Rancher: master

Steps:
-  add a single node cluster
- go to tools -> notofies to add Slcak as the notifer 
- go to tools -> alerts 
- at the bottom, there is a group called `A set of alerts for node`
<img width="1435" alt="screenshot" src="https://user-images.githubusercontent.com/6218999/57420200-5c20fe80-71bb-11e9-9735-98a477738bc9.png">
- edit the group, scroll down to the bottom, add Slack as the reciver, then click on save

<img width="1531" alt="screenshot" src="https://user-images.githubusercontent.com/6218999/57420267-a4d8b780-71bb-11e9-9f2b-cf07c06b3f3c.png">

See the following error
<img width="1555" alt="screenshot" src="https://user-images.githubusercontent.com/6218999/57420272-ad30f280-71bb-11e9-9fb4-ce87654ad91e.png">

In this test, I did not enable the cluster monitoring at all and used the built-in alerts

**Comment by jiaqiluo** [comment]
per to the conversation with @loganhz the behavior I described in my previous comment is expected.   

As a conclusion, the bug fix is validated and the issue is closed.

**Comment by vincent927** [comment]
> @doog33k @materemias @hmonadjem @vincent927
> 
> We will fix it in 2.2.2.
> 
> If you hit this issue in 2.2.0 or 2.2.1, currently the workaround is not easy.
> But it can get your cluster monitoring working again.
> 
> Workaround in 2.2.0 and 2.2.1 if you manually delete `cattle-prometheus` namespace.
> 
> 1. Disable cluster monitoring
> 2. Go to System project -> Apps
> 3. Remove `cluster-monitoring` and `monitoring-operator` apps if the apps cannot be deleted.
> 4. docker exec into your rancher server container
> 5. `kubectl get cluster` to find your cluster id (You might need to install vim in the container)
> 6. `kubectl edit cluster ${cluster_id}`
> 7. Delete the `PrometheusOperatorDeployed` condition and save the change. (Remove the three lines in the screenshort)
> 8. Enable the cluster monitoring again
>    ![image](https://user-images.githubusercontent.com/21168270/55466854-a0583680-5632-11e9-9faa-e5fc2f6b8310.png)

@loganhz  hi ,I tried to solve the similar problem of cluster alerting according to your method, but it did not succeed. Is there any way to solve it? There is nothing in the system cluster alerting app now.

![image](https://user-images.githubusercontent.com/10399425/59397838-a1d77500-8dc0-11e9-9820-456cfd72804c.png)

**Comment by diogo-as** [comment]
Thanks @loganhz for your answer.
I tried to use your workarround on rancher v2.2.2 and works fine!
